File: yukicoder/tyama_yukicoder472.py
#!/usr/bin/python
from functools import sys
import sys
sys.setrecursionlimit(1000000)
memo={}
def dfs(v,d,r):
	if d==len(v): return 0
	if (d,r) not in memo: memo[(d,r)]=reduce(lambda s,i: min(s,dfs(v,d+1,r-i)+v[d][i]), range(min(r,3)+1), 1<<30)
	return memo[(d,r)]

n,k=map(int,sys.stdin.readline().split())
v=[list(map(int,sys.stdin.readline().split()))+[1] for _ in range(n)]
print(dfs(v,0,k)/float(n))

File: yukicoder/tyama_yukicoder472.rb
#!/usr/bin/ruby
$memo={}
def dfs(v,d,r)
	return 0 if d==v.size
	$memo[[d,r]]||=(0..[r,3].min).reduce(1<<30){|s,i|
		[s,dfs(v,d+1,r-i)+v[d][i]].min
	}
end
n,k=gets.split.map(&:to_i)
v=$<.map{|e|e.split.map(&:to_i)+[1]}
p dfs(v,0,k)/n.to_f
